src/components/shapes/attributes.js

```diff
@@ -220,7 +220,28 @@ module.exports = templatedArray('shape', {
             'See `type` and `ysizemode` for more info.'
         ].join(' ')
     },
-
+    x_shift: {
+        valType: 'number',
+        dflt: 0,
+        min: -0.5,
+        max: 0.5,
+        editType: 'calc',
+        description: [
+            'Only relevant if xref is a (multi-)category axes. Shifts x0 and x1 by a fraction of',
+            'the reference unit.'
+        ].join(' ')
+    },
+    y_shift: {
+        valType: 'number',
+        dflt: 0,
+        min: -0.5,
+        max: 0.5,
+        editType: 'calc',
+        description: [
+            'Only relevant if yref is a (multi-)category axes. Shifts y0 and y1 by a fraction of',
+            'the reference unit.'
+        ].join(' ')
+    },
```

Comment:
Please have a look at `text_on_shapes_basic` mock where you could draw new shapes as described by the `json` spec on the graph.
As per my comment https://github.com/plotly/plotly.js/pull/7005#issuecomment-2142381614
I am still not convinced that these new attributes belong to shape and selection not the axis.

Reply:
Yeah seems weird for newshape. It would need to be factored into the calculation for converting what you draw to coordinates, so what’s the purpose? I’d suggest omitting it until we see a clear use case.

Reply:
I am thinking here that instead of just shifting values, what users want is to round them to categories (i.e. integers) and then use these offsets to declare the start and end. A feature like `snap to grid`. 
This would make selection on categories easier and could have application in other axis types too.

Reply:
Hi @archmoj is it possible to postpone this to a later PR? I didn't anticipate this feature and haven't looked at newshape/newselection before, so I'm not quite sure how to approach this feature atm.